frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts

```diff
@@ -36,60 +41,83 @@ export class MetricsComponent implements OnInit, OnDestroy, AfterViewInit {
 
   selectedMetricIndex = null;
   metricFilterOptions = [METRIC_SEARCH_KEY.ALL, METRIC_SEARCH_KEY.NAME, METRIC_SEARCH_KEY.CONTEXT];
   selectedMetricFilterOption = METRIC_SEARCH_KEY.ALL;
   contextOptions: string[] = [];
   searchValue: string;
 
-  constructor(private analysisService: AnalysisService, private authService: AuthService, private dialog: MatDialog) {}
+  constructor(private analysisService: AnalysisService, private authService: AuthService, private dialog: MatDialog) {
+    // Initialize NestedTreeControl with custom getChildren method for lazy loading
+    this.nestedTreeControl = new NestedTreeControl<MetricUnit | LazyLoadingMetric>(this.getChildren);
+  }
 
   ngOnInit() {
     this.permissionSub = this.authService.userPermissions$.subscribe((permission) => {
       this.permissions = permission;
     });
 
     this.allMetricsSub = this.analysisService.allMetrics$.subscribe((metrics) => {
-      this.allMetrics = new MatTableDataSource();
-      this.allMetrics.data = metrics.map((item) => {
-        this.insertNodeIndex = 0;
-        return this.insertNode(item);
-      });
+      this.allMetrics = new MatTableDataSource<LazyLoadingMetric>();
+      // Process metrics data to prepare for lazy loading
+      this.allMetrics.data = this.processMetricsData(metrics);
       this.extractContext(this.allMetrics.data);
     });
 
     this.applyFilter(this.searchValue);
   }
 
-  hasNestedChild = (_: number, nodeData: IMetricUnit) => !!nodeData.children && nodeData.children.length > 0;
+  // Determine if a node has nested children (either loadable or already loaded)
+  hasNestedChild = (_: number, node: LazyLoadingMetric): boolean =>
+    !!node.loadChildren || (node.children && node.children.length > 0);
 
-  insertNode(metrics: any): MetricUnit {
-    if (!metrics.children.length) {
-      const data = { id: this.insertNodeIndex, ...metrics };
-      this.insertNodeIndex += 1;
-      return data;
-    }
-    metrics = {
-      id: this.insertNodeIndex,
-      ...metrics,
-      children: metrics.children.map((data) => {
-        this.insertNodeIndex += 1;
-        data = this.insertNode(data);
-        return data;
-      }),
+  // Process a single metric node, setting up lazy loading for its children
+  private insertNode(metric: IMetricUnit): LazyLoadingMetric {
+    const processedMetric: LazyLoadingMetric = {
+      ...metric,
+      children: [],
     };
-    return metrics;
+
+    // If the node has children, create a loadChildren function instead of processing them immediately
+    if (metric.children && metric.children.length > 0) {
+      processedMetric.loadChildren = () => Promise.resolve(metric.children?.map((child) => this.insertNode(child)));
+    }
+
+    return processedMetric;
+  }
+
+  // Custom method to get children for a node, supporting lazy loading
+  private getChildren = (node: LazyLoadingMetric): Observable<LazyLoadingMetric[]> => {
```

Comment:
then this can be smooshed down to:

```
private getChildren = (node: LazyLoadingMetric): Observable<LazyLoadingMetric[]> => {
    if (node.loadChildren) {
      return node.loadChildren().pipe(tap((children) => (node.children = children)));
    }
    return of(node.children || []);
  };
  ```